src/rpc/pd/timestamp.rs

```diff
@@ -3,226 +3,185 @@
-                        Ok(())
-                    }
-                    Err(e) => {
-                        error!("failed to send tso requests: {:?}", e);
-                        Err(())
-                    }
-                })
-                .compat(),
-        );
-
-        handle.spawn(
-            rx.try_for_each(move |resp| {
-                let reactor = &mut r.write().unwrap();
-                let tso_pending = reactor.tso_pending.take().unwrap();
-                reactor.schedule(Task::Response(tso_pending, resp));
-                if !reactor.tso_batch.is_empty() {
-                    // Schedule another tso_batch of request
-                    reactor.schedule(Task::Request);
+        let mut count = 0;
+        while ctx.pending_queue.len() < MAX_PENDING_COUNT {
+            match self.result_sender_rx.as_mut().poll_next(cx) {
+                Poll::Ready(Some(sender)) => {
+                    ctx.pending_queue.push_back(sender);
+                    count += 1;
                 }
-                future::ready(Ok(()))
-            })
-            .map_err(|e| panic!("unexpected error: {:?}", e))
-            .compat(),
-        );
-    }
-
-    fn tso_request(&mut self, cluster_id: u64) {
-        let mut tso_batch = self.tso_buffer.take().unwrap();
-        tso_batch.extend(self.tso_batch.drain(..));
-        let mut request = pd_request!(cluster_id, pdpb::TsoRequest);
-        let batch_size = tso_batch.len();
-        observe_tso_batch(batch_size);
-        request.count = batch_size as u32;
-        self.tso_pending = Some(tso_batch);
-        self.tso_tx
-            .try_send(request)
-            .expect("channel can never be full");
-    }
+                Poll::Ready(None) => return Poll::Ready(None),
+                Poll::Pending => break,
+            }
+        }
 
-    fn tso_response(&mut self, mut requests: Vec<TsoChannel>, response: pdpb::TsoResponse) {
-        let timestamp = response.get_timestamp();
-        for (offset, request) in requests.drain(..).enumerate() {
-            request
-                .send(Timestamp {
-                    physical: timestamp.physical,
-                    logical: timestamp.logical + offset as i64,
-                })
-                .unwrap();
+        if count > 0 {
+            let req = TsoRequest {
+                header: Some(RequestHeader {
+                    cluster_id: ctx.cluster_id,
+                }),
+                count,
+            };
+            let write_flags = WriteFlags::default().buffer_hint(false);
+            Poll::Ready(Some((req, write_flags)))
+        } else {
+            Poll::Pending
         }
-        self.tso_buffer = Some(requests);
     }
+}
 
-    fn schedule(&self, task: Task) {
-        self.task_tx
-            .as_ref()
-            .unwrap()
-            .unbounded_send(task)
-            .expect("unbounded send should never fail");
-    }
+struct TsoContext {
```

Comment:
Can you find a better name here too? `Context` is not very descriptive.

Reply:
Sorry but I cannot think of an appropriate name. Could you help me?
The struct is just what I want to share in both the `send_request` future and the `receive_and_handle_response` future. The `send_request` future stores its waker in it and pushes timestamp senders into the queue. The `receive_and_handle_response` future dequeues the queue and wake up the former future.

Reply:
Right. I have the same feeling too. It is not something that can be clearly described. It is created according to what we need.
They are something that both futures need to access and mutate. If we don't have such a struct, we have to make two `Rc<RefCell>`s (one for setting the waker, one for mutating the queue).
However, I will try and see if I can do an ideal refactoring.